File: shared/src/main/scala/org/mulesoft/language/server/common/reconciler/Reconciler.scala
package org.mulesoft.language.server.common.reconciler;

import org.mulesoft.language.common.logger.ILogger

import scala.collection.mutable.ListBuffer
import scala.concurrent.{Future, Promise}
import scala.concurrent.ExecutionContext.Implicits.global
import scala.util.{Failure, Success}
import java.util.{Timer, TimerTask}

class Reconciler(logger: ILogger, timeout: Int, setTimeout: Function2[Function0[Unit], Int, Unit] = (task: Function0[Unit], timeout: Int) => {
	new Timer().schedule(new TimerTask {
		def run = task();
	}, timeout);
}) {
	private var waitingList: ListBuffer[Runnable[Any]] = ListBuffer[Runnable[Any]]();
	private var runningList: ListBuffer[Runnable[Any]] = ListBuffer[Runnable[Any]]();
	
	def shedule[ResultType](runnable: Runnable[ResultType]): Promise[ResultType] = {
		var result = Promise[ResultType]();
		
		addToWaitingList(runnable);
		
		setTimeout(() => {
			if(runnable.isCanceled()) {
				removeFromWaitingList(runnable);
			} else {
				findConflictingInRunningList(runnable) match {
					case Some(found) => shedule(runnable);
					
					case _ => {
						removeFromWaitingList(runnable);
						
						addToRunningList(runnable);
						
						run(runnable).future.andThen {
							case Success(value) => result.success(value);
							case Failure(error) => result.failure(error);
						};
					}
				};
			}
		}, timeout);
		
		result;
	}
	
	private def run[ResultType](runnable: Runnable[ResultType]): Promise[ResultType] = {
		var result = Promise[ResultType]();
		
		runnable.run().future.andThen {
			case Success(success) => {
				removeFromRunningList(runnable);
				
				result.success(success);
			}
			
			case Failure(error) => {
				removeFromRunningList(runnable);
				
				result.failure(error);
			}
		};
		
		result;
	}
	
	private def addToWaitingList[ResultType](runnable: Runnable[ResultType]) {
		waitingList = waitingList.filter(current => {
			var conflicts = runnable.conflicts(current);
			
			if(conflicts) {
				current.cancel();
			}
			
			!conflicts;
		});
		
		waitingList += runnable.asInstanceOf[Runnable[Any]];
	}
	
	private def addToRunningList[ResultType](runnable: Runnable[ResultType]) {
		runningList += runnable.asInstanceOf[Runnable[Any]];
	}
	
	private def removeFromWaitingList[ResultType](runnable: Runnable[ResultType]) {
		
		waitingList -= runnable.asInstanceOf[Runnable[Any]];
	}
	
	private def removeFromRunningList[ResultType](runnable: Runnable[ResultType]) {
		runningList -= runnable.asInstanceOf[Runnable[Any]];
	}
	
	private def findConflictingInRunningList[ResultType](runnable: Runnable[ResultType]): Option[Runnable[ResultType]] = runningList.find(runnable.conflicts(_)).asInstanceOf[Option[Runnable[ResultType]]];
}

File: js/static/raml-language-server/src/common/reconciler.ts
import {
    ILogger
} from "./logger";

import PromiseConstructor = require("promise-polyfill");
if (typeof Promise === "undefined" && typeof window !== "undefined") {
    (window as any).Promise = PromiseConstructor;
}

export interface Runnable<ResultType> {

    /**
     * Performs the actual business logics.
     * Should resolve the promise when finished.
     */
    run(): Promise<ResultType>;

    // Commented out as we do not allow to run parsing synhronously any more due to the connection,
    // which provides file system information does this only asynchronously
    // /**
    //  * Performs the actual business logics synchronously.
    //  */
    // runSynchronously() : ResultType;

    /**
     * Whether two runnable conflict with each other.
     * Must work fast as its called often.
     * @param other
     */
    conflicts(other: Runnable<any>): boolean;

    /**
     * Cancels the runnable. run() method should do nothing if launched later,
     * if cancel is called during the run() method execution, run() should stop as soon as it can.
     */
    cancel(): void;

    /**
     * Whether cancel() method was called at least once.
     */
    isCanceled(): boolean;
}

export class Reconciler {

    private waitingList: Runnable<any>[] = [];
    private runningList: Runnable<any>[] = [];

    constructor(private logger: ILogger, private timeout: number) {
    }

    public schedule<ResultType>(runnable: Runnable<ResultType>): Promise<ResultType> {
        this.addToWaitingList(runnable);

        return new Promise((resolve: (value?: ResultType) => void, reject: (error?: any) => void) => {

            setTimeout(() => {

                this.logger.debugDetail("Time came to execute " + runnable.toString(),
                    "Reconciler", "schedule");

                if (runnable.isCanceled()) {
                    this.logger.debugDetail("Runnable " + runnable.toString() + " is cancelled, doing nothing",
                        "Reconciler", "schedule");

                    this.removeFromWaitingList(runnable);
                    return;
                }

                const currentlyRunning = this.findConflictingInRunningList(runnable);
                if (currentlyRunning) {
                    // TODO add an additional short timeout parameter to launch the reschedule
                    // at the finish of the currently running task for a short time after it.

                    this.logger.debugDetail("Conflicting to " + runnable.toString()
                        + " is found in the running list: " + currentlyRunning.toString()
                        + " rescheduling current one.",
                        "Reconciler", "schedule");

                    this.schedule(runnable);
                    return;
                }

                this.removeFromWaitingList(runnable);
                this.addToRunningList(runnable);

                this.logger.debugDetail("Executing " + runnable.toString(),
                    "Reconciler", "schedule");

                this.run(runnable).then(
                    (result) => { resolve(result); },
                    (error) => { reject(error); }
                );

            }, this.timeout);

        });
    }

    private run<ResultType>(runnable: Runnable<ResultType>): Promise<ResultType> {

        return runnable.run().then(

            (result): ResultType => {
                this.removeFromRunningList(runnable);
                return result;
            },
            (error): ResultType => {
                this.removeFromRunningList(runnable);
                throw error;
            }

        );
    }

    /**
     * Adds item to waiting list and removes anything currently in the list,
     * which conflicts with the new item.
     * @param runnable
     */
    private addToWaitingList<ResultType>(runnable: Runnable<ResultType>) {
        this.logger.debugDetail("Adding runnable " + runnable.toString() + " to waiting list",
            "Reconciler", "addToWaitingList");

        this.waitingList = this.waitingList.filter((current) => {

            this.logger.debugDetail("Comparing existing runnable " + current.toString() +
                " to the new " + runnable.toString(),
                "Reconciler", "addToWaitingList");

            const conflicts = runnable.conflicts(current);
            if (conflicts) {
                this.logger.debugDetail("Runnables are conflicting, canceling existing one",
                    "Reconciler", "addToWaitingList");

                current.cancel();
            }

            return !conflicts;
        });

        this.waitingList.push(runnable);
    }

    /**
     * Removes runnable from the list of running ones.
     * @param runnable
     */
    private removeFromWaitingList<ResultType>(runnable: Runnable<ResultType>) {
        this.logger.debugDetail("Removing " + runnable.toString()
            + " from waiting list",
            "Reconciler", "removeFromWaitingList");

        const index = this.waitingList.indexOf(runnable);
        if (index !== -1) {
            this.waitingList.splice(index, 1);
        }
    }

    /**
     * Adds runnable to the list of running ones.
     * @param runnable
     */
    private addToRunningList<ResultType>(runnable: Runnable<ResultType>) {
        this.logger.debugDetail("Adding " + runnable.toString()
            + " to running list",
            "Reconciler", "removeFromWaitingList");

        this.runningList.push(runnable);
    }

    /**
     * Removes runnable from the list of running ones.
     * @param runnable
     */
    private removeFromRunningList<ResultType>(runnable: Runnable<ResultType>) {
        this.logger.debugDetail("Removing " + runnable.toString()
            + " from running list",
            "Reconciler", "removeFromWaitingList");

        const index = this.runningList.indexOf(runnable);
        if (index !== -1) {
            this.runningList.splice(index, 1);
        }
    }

    /**
     * Finds the first conflicting runnable in the current list.
     * @param runnable
     * @returns {any}
     */
    private findConflictingInRunningList<ResultType>(runnable: Runnable<ResultType>): Runnable<ResultType> {
        for (const current of this.runningList) {
            if (runnable.conflicts(current)) {
                return current;
            }
        }

        return null;
    }
}
